Add more Dune tips to hacking doc. (#14227)
coq/coq: dev/doc/README.md

recommended to use a recent OCaml version, which they can get through
opam or Nix, in particular.

## Setting up Dune caching

You will get faster rebuilds if you enable Dune caching. This is true
in all cases, but even more so when using the targets in
`Makefile.dune` (see below).

To set up Dune caching, you should add a Dune configuration file
(generally located in `~/.config/dune/config` with the following
content:

```
(lang dune 2.0)
(cache enabled)
```

> 

Reviewer: Should we warn about disk use ?

Author: OK, done.

recommended to use a recent OCaml version, which they can get through
opam or Nix, in particular.

## Configuring Dune caching and default verbosity

> There are several configuration settings that you can control globally
> with a Dune configuration file (see `man dune-config` to learn
> more). This file is generally located in `~/.config/dune/config` and
> should start with the version of the Dune language used, e.g.:

Reviewer 2: ```suggestion
There are several configuration settings that you can control globally
by creating a Dune configuration file (see `man dune-config` to learn
more). Dune uses the file `~/.config/dune/config` by default.  The file
should start with the version of the Dune language used, e.g.:
```

I didn't have such a file.  If users should always have such a file, please explain how it gets created.  If so, it would be nice if something would automatically create it.

Reviewer: That is unfortunately system-specific. In normal operation the file is not needed.

Reviewer: That == the path

```
(lang dune 2.0)
```

> 

Reviewer 2: ```suggestion

`(lang dune 2.0)` is the language version setting to use for any Dune software version `2.*.*`.
```

But https://dune.readthedocs.io/en/stable/dune-files.html shows `(lang dune 2.8)`.  What's right?

Reviewer: Both, dune's build language is versioned, so pick the version you want / need.

Author: I'll leave this one unchanged. IMHO if you want to know all the details, you can look at the Dune doc.

- You will get faster rebuilds if you enable Dune caching. This is
  true in all cases, but even more so when using the targets in
  `Makefile.dune` (see below).

  To set up Dune caching, you should append the following line to your
  Dune configuration file:

  ```
  (cache enabled)
  ```

  Note that by default, Dune caching will use up to 10GB of disk size.
  See the [official documentation](https://dune.readthedocs.io/en/stable/caching.html#on-disk-size)

>   to learn how to change the default.

Reviewer 2: 10 GB seems rather piggy considering that the build directory is only about 1GB.  I wonder how well dune handles building in multiple source trees (serially or in parallel).

Changing the default is so simple IIUC you might as well mention it here, i.e. `(cache-trim-size 999MB)`.

The webpage link here and the man page mentioned on line 13 could go at the very beginning of the section, I think.

Reviewer: For some people the cache is overkill, others have cache sizes in the TiB , so indeed YMMV here. 10 GiB is a good default for the regular OCaml developer.

> I wonder how well dune handles building in multiple source trees (serially or in parallel).

Not sure what you mean by this, I'll be happy to answer over Zulip tho.

Reviewer: Note that the cache had some known issues on windows, I understand it has much improved lately, but TTBOMK it is still experimental on that platform.

Reviewer 2: FWIW, I use WSL2 (Ubuntu VM on Windows), not native Windows.

- Dune is not very verbose by default. If you want to change the
  behavior to a more verbose one, you may append the following line to
  your Dune configuration file:

  ```
  (display short)
  ```

> 

Reviewer 2: FWIW the Dune doc doesn't mention how to put a comment in the file.  What is the convention?

Reviewer: LISP syntax, that is to say `; comment`

    # to build some test project
```

When running the commands above, you may set `DUNEOPT=--display=short`
for a more verbose build.

> 

Reviewer: Or alternatively update `.config/dune/` to set your display pref.

Author: I didn't find anywhere in the Dune documentation documenting this. The only place that I know of that talks about this file is https://dune.readthedocs.io/en/stable/caching.html#configuration

Reviewer: Seems to be only in `man dune-config`, indeed the last rewrite of the Dune doc was not very convincing IMO.

Reviewer: Removed the Coq section, forgot to add a section for the config file, etc... I think that section was there before.

Author: OK thanks for the tip about `man dune-config`.
